In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")


# Exploratory Analysis & PreProcessing

In [2]:
df_train = pd.read_csv('data.csv')

In [3]:
df_train.shape # Shape of our training set

(10000, 3)

In [4]:
df_train.head()

address   latitude  longitude
0  JAIPUR,H.NO.- 408, LAVENDER MANGALAM, ANAND NA...  71.052021  11.396546
1                       C-341-A,Malviya nagar,302017  70.516921   9.855150
2  J-3A, Khandal Hostel, Jhalana Doongri,,Jaipur,...  70.430068   9.222819
3  846  Rani Sati Nagar  Janpath  Lane No. 10  Aj...  71.333194   8.762032
4  12/132 , sector 12,girdhar marg, malviya nagar...  70.490080  10.209948

In [5]:
df_train['Comma_delimitted_address']=df_train['address'].str.split(',') #Comma delimitting string

In [6]:
df_train['zip']=df_train['Comma_delimitted_address']

In [7]:
for n in range(10000):
    df_train['zip'][n]=df_train['Comma_delimitted_address'][n][-1] #Extracting Zip Codes from address
    df_train['address'].iloc[n]=df_train['address'].iloc[n].lower() #Converting address to lowercase

In [8]:
df=df_train.groupby('zip').describe()
df ## Simple exploratory analysis giving us insights about various zipcodes associated with our addresses 

latitude                                                        \
          count       mean       std        min        25%        50%   
zip                                                                     
302001    227.0  70.563646  0.169794  70.145060  70.457821  70.508820   
302002    200.0  70.244976  0.116888  70.041760  70.156119  70.262491   
302003    136.0  70.244915  0.111343  70.014251  70.138482  70.247168   
302004    576.0  70.243602  0.084020  70.067955  70.184189  70.238455   
302005     51.0  70.705687  0.037339  70.615241  70.680351  70.705253   
302006    326.0  70.889371  0.116174  70.635735  70.806431  70.870130   
302007      2.0  70.591596  0.451391  70.272414  70.432005  70.591596   
302011      7.0  70.660223  0.321424  70.038364  70.595341  70.742748   
302012    959.0  71.369186  0.217888  70.756466  71.233200  71.337373   
302013    189.0  70.980195  0.142525  70.649209  70.908318  70.985594   
302014      1.0  70.731996       NaN  70.731996  70.731996  70.731996   
302015    684.0  70.713571  0.110242  70.479366  70.631663  70.710130   
302016    446.0  70.694441  0.080918  70.535198  70.643741  70.686271   
302017    913.0  70.361840  0.226399  69.554652  70.265764  70.448200   
302018    305.0  70.773002  0.133456  70.465360  70.675310  70.735361   
302019    661.0  71.091905  0.146362  69.802296  71.006916  71.069212   
302020   1557.0  71.073199  0.115720  70.597426  70.984859  71.056859   
302021    606.0  71.383509  0.126748  70.935663  71.301641  71.383076   
302022    150.0  70.206939  0.179548  69.841053  70.103274  70.221953   
302023    271.0  70.898166  0.084067  70.691177  70.849518  70.895567   
302024     24.0  71.506972  0.098861  71.287250  71.463084  71.494451   
302025     27.0  70.004842  0.186892  69.817614  69.855964  69.919331   
302026     99.0  72.000056  0.356724  71.553839  71.865244  71.905529   
302027     32.0  69.732550  0.143080  69.128058  69.669573  69.770409   
302028     89.0  69.756957  0.191417  69.463361  69.521730  69.799290   
302029    378.0  70.911547  0.172049  70.160353  70.841417  70.892845   
302030     13.0  70.356762  0.232458  70.112289  70.114871  70.408631   
302031     38.0  69.437709  0.104707  69.287338  69.366659  69.409751   
302032      5.0  71.100965  0.098228  70.947195  71.084883  71.128206   
302033    173.0  70.417916  0.168052  70.051369  70.295413  70.426061   
302034    124.0  71.628546  0.108385  71.231447  71.547333  71.610661   
302036      4.0  69.745325  0.081679  69.683669  69.685610  69.719903   
302039    528.0  70.985498  0.119031  70.713037  70.893557  70.982277   
302040      5.0  71.391839  0.081312  71.249772  71.397139  71.437373   
303002      1.0  71.257427       NaN  71.257427  71.257427  71.257427   
303007    165.0  73.546308  0.103785  73.311057  73.484809  73.485374   
303012     25.0  69.658525  0.219839  69.077415  69.540059  69.560444   
303905      3.0  69.718225  0.001631  69.717065  69.717293  69.717522   

                             longitude                                  \
              75%        max     count       mean       std        min   
zip                                                                      
302001  70.678291  71.278342     227.0   7.499307  0.453884   5.733055   
302002  70.321249  71.077559     200.0   6.794372  0.326928   5.963971   
302003  70.323494  70.868688     136.0   7.535019  0.204884   7.259255   
302004  70.292628  70.724412     576.0   8.368404  0.268464   7.666205   
302005  70.733307  70.785151      51.0   8.491171  0.204339   7.344134   
302006  70.966213  71.530073     326.0   8.053026  0.376979   7.147274   
302007  70.751187  70.910777       2.0   8.097628  0.132911   8.003646   
302011  70.791978  71.065814       7.0  11.022410  0.364164  10.500514   
302012  71.435410  72.515009     959.0   6.497777  0.542624   3.800507   
302013  71.075931  71.339939     189.0   5.092910  0.448873   3.952547   
302014  70.731996  70.731996       1.0   9.339649  

# Tokenizing the addresses

In [9]:
df_train['tokens'] = df_train['Comma_delimitted_address']
import nltk
for n in range(10000):
    tokens = nltk.word_tokenize(df_train['address'][n])
    exceptions = [',','jaipur','rajasthan' ,'near', 'road', 'apartment', 'adjacent','sector']
    #^ Customized stopwords that will have high colinearity with a wide range of addresses 
    #These words if not removed can skew Jaccard similarity between two unrelated addresses significantly
    #My exceptions/stopwords list by no means is exhaustive its just an attempt to showcase what needs to be done
    df_train['tokens'][n]=set(tokens).difference(exceptions)
    #Removing our custom set of stopwords

In [10]:
df_train.head()

address   latitude  longitude  \
0  jaipur,h.no.- 408, lavender mangalam, anand na...  71.052021  11.396546   
1                       c-341-a,malviya nagar,302017  70.516921   9.855150   
2  j-3a, khandal hostel, jhalana doongri,,jaipur,...  70.430068   9.222819   
3  846  rani sati nagar  janpath  lane no. 10  aj...  71.333194   8.762032   
4  12/132 , sector 12,girdhar marg, malviya nagar...  70.490080  10.209948   

                            Comma_delimitted_address     zip  \
0  [JAIPUR, H.NO.- 408,  LAVENDER MANGALAM,  ANAN...  302029   
1                   [C-341-A, Malviya nagar, 302017]  302017   
2  [J-3A,  Khandal Hostel,  Jhalana Doongri, , Ja...  302004   
3  [846  Rani Sati Nagar  Janpath  Lane No. 10  A...  302019   
4  [12/132 ,  sector 12, girdhar marg,  malviya n...  302017   

                                              tokens  
0  {mangalam, 408, 302029, railway, -, o-302029, ...  
1                   {nagar,302017, malviya, c-341-a}  
2  {doongri, khandal, j-3a, ,jaipur,302004, hoste...  
3  {lane, janpath, india,302019,302019, ., nagar,...  
4  {girdhar, marg, nagar,302017, 12, 12/132, malv...

# Calculating Jaccard Score within the same zipcode subgroup (Test Case used here is Zip Code 302029)

** Jaccard Score is a way to measure the similarity of two strings, its can be summarized as the  Intersection over Union of the sets at play **


** Higher score corresponds to higher similarity within two strings**

In [11]:
df_temp= df_train[df_train['zip']=='302029'] #Test Case of Sub Group ZipCode = 302029
df_temp['similarity']= df_temp['zip']

In [12]:
for i in range(len(df_temp)): #Calculating Jaccard Scores for element '0' w.r.t. other elements with the same zipcode
    intersection = df_temp['tokens'].iloc[0].intersection(df_temp['tokens'].iloc[i])
    jaccard_score= float(len(intersection))/((len(df_temp.tokens.iloc[0])) + len(df_temp.tokens.iloc[i]) - len(intersection))
    df_temp['similarity'].iloc[i] = jaccard_score

In [13]:
df_temp.sort_values(by='similarity', ascending=False)

address   latitude  longitude  \
0     jaipur,h.no.- 408, lavender mangalam, anand na...  71.052021  11.396546   
5201  jaipur,h.no.- 408, lavender mangalam, anand na...  71.051969  11.398205   
7369  rose g-8 , mangalam aananda, opposite sanganer...  71.058382  11.359776   
7974  rose g-8 , mangalam aananda, opposite sanganer...  71.062842  11.351647   
8244  41 sunder nagar sanganer railway station, ke s...  70.999621  11.356692   
5702  flat 306, block iris, mangalam ananda,near san...  71.045985  11.359348   
2250  bright cotton b4, khatri nagar, near sanganer ...  70.892179  10.845294   
8660  bright cotton b4, khatri nagar, near sanganer ...  70.891893  10.846765   
6491  bright cotton b4, khatri nagar, near sanganer ...  70.892897  10.848949   
4041  bright cotton b4, khatri nagar, near sanganer ...  70.892786  10.848944   
4248  bright cotton b4, khatri nagar, near sanganer ...  70.892586  10.847497   
4404  bright cotton b4, khatri nagar, near sanganer ...  70.892897  10.848949   
8816  26 dev nager sanganer railway station jaipur ,...  70.985133  11.441609   
7971  lavender 207 mangalam aananda sanganer,mansaro...  71.051058  11.394790   
5838  near railway station sanganer kankho ki dhani|...  70.967692  11.427949   
1653  adinath traders shop no.40 railway station roa...  70.944240  11.258171   
7226  gulab nagar sanaganer railway station ke saman...  70.971156  11.187452   
4306  gulab nagar sanaganer railway station ke saman...  70.971156  11.187452   
1137  bright cotton 2nd floor, shop no. 6 krishna pl...  70.844079  11.210822   
1438  108 jasmine block manglam ananda #near sangane...  71.050630  11.357928   
2112  h. no -6 heerawat nagar , sanganer , near opp ...  70.951640  11.328842   
1936           sanganer,sanganer railway station,302029  70.987018  11.138759   
7227  khasra no.725 prem nagar ke pass haziawala, sa...  70.999661  11.956528   
6476  khasra no.725 prem nagar ke pass haziawala, sa...  71.058147  11.846815   
7110  h no 17 jagannath puri colony mal ki dhani san...  70.846946  11.821218   
7209  khasra no.725 prem nagar ke pass haziawala, sa...  71.058983  11.839512   
2076  104, lavendar,mangalam's aananda, sanganer,in ...  71.052389  11.403921   
4259  khasra no.725 prem nagar ke pass haziawala, sa...  71.059521  11.834112   
3164  khasra no.725 prem nagar ke pass haziawala, sa...  71.060093  11.833522   
7949  khasra no.725 prem nagar ke pass haziawala, sa...  71.058780  11.834718   
...                                                 ...        ...        ...   
5070      35 sunita colony rampura road,sanganer,302029  71.064642  11.194208   
5019  0 \n malpura get roadwesh bus stend sannganer ...  70.871393  11.404492   
8469  12s olive homes,kanka ki dhani , iskon road, 2...  71.185278  11.156432   
8323    saipura stand ston age ke samne sanganer,302029  71.019738  12.783726   
3526  hiya enterprises_x000d_\n121 shiv marg ashind ...  70.860467  11.065567   
3533  opp.  tulsi centre \n near cts bus stand \n sa...  70.828600  11.319705   
8277  shop no-39,laxmi vihar colony,rampura road, , ...  71.153864  11.104908   
3539        73,kailashpuri rampura road,sanganer,302029  71.169907  11.188527   
3607  7/28 ,jain mohalla c.t.s bus stand , sanganer,...  70.822765  11.322679   
3740  near by shiv shakti tower, mahindra world city...  72.633573  12.039684   
3846  shop no-39,laxmi vihar colony,rampura road, , ...  71.152507  11.095558   
8126  phed office campus,  city bus stand,sanganer,3...  70.847418  11.317058   
3921  shree nath chamber, opp. pingrapol gosala,main...  70.602453  11.479138   
3958  c-39a,prem colony,taaron ki koont ,tonk road,3...  70.813246  10.534749   
8070  35, sunita colony, rampura road,near vishakha ...  71.064051  11.191379   
4058  ashish tredrs,  shop no. 8 ,narania hp gas ke ...  70.946771  11.276525   
4188  shakambhari kohinoor residency,diamond-804, ra...  70.997255  11.234813   
4233         112gha radha vallabh marg ,sanganer,302029  70.790426  11.286334  

# Calculating Jaccard Score for test case element '0' w.r.t. to all other addresses , and calculating relative distance from each of these points 

**Distance metric is calculate here assuming long/lat cordinates to be analogous to cordintaes in a Euclidean Space. This is done only because these lat/longitudes are encoded and dont refelect true positions on the globe so a Euclidean system should sufficiently capture the crux of the relationship (Nearness/Furtherness)**

In [14]:
df_copy=df_train
df_copy['similarity'] = df_copy['zip']
df_copy['distance'] = df_copy['zip']
import math  
for i in range(len(df_copy)):
    intersection = df_copy['tokens'].iloc[0].intersection(df_copy['tokens'].iloc[i])
    jaccard_score= float(len(intersection))/((len(df_copy.tokens.iloc[0])) + len(df_copy.tokens.iloc[i]) - len(intersection))
    df_copy['similarity'].iloc[i] = jaccard_score 
    dist = math.sqrt((df_copy['latitude'].iloc[i] - df_copy['latitude'].iloc[0])**2 + (df_copy['longitude'].iloc[i] - df_copy['longitude'].iloc[0])**2)  
    df_copy['distance'].iloc[i] = dist  
    

## Underneath I have sorted the findings using similarity scores , and distance metric 

In [15]:
df_copy.sort_values(by='similarity', ascending=False)

address   latitude  longitude  \
0     jaipur,h.no.- 408, lavender mangalam, anand na...  71.052021  11.396546   
5201  jaipur,h.no.- 408, lavender mangalam, anand na...  71.051969  11.398205   
7974  rose g-8 , mangalam aananda, opposite sanganer...  71.062842  11.351647   
7369  rose g-8 , mangalam aananda, opposite sanganer...  71.058382  11.359776   
8244  41 sunder nagar sanganer railway station, ke s...  70.999621  11.356692   
5702  flat 306, block iris, mangalam ananda,near san...  71.045985  11.359348   
6491  bright cotton b4, khatri nagar, near sanganer ...  70.892897  10.848949   
8660  bright cotton b4, khatri nagar, near sanganer ...  70.891893  10.846765   
4248  bright cotton b4, khatri nagar, near sanganer ...  70.892586  10.847497   
4041  bright cotton b4, khatri nagar, near sanganer ...  70.892786  10.848944   
4404  bright cotton b4, khatri nagar, near sanganer ...  70.892897  10.848949   
2250  bright cotton b4, khatri nagar, near sanganer ...  70.892179  10.845294   
1573   91 , gulab nagar railway station sanganer,302020  70.973082  11.315466   
5838  near railway station sanganer kankho ki dhani|...  70.967692  11.427949   
9338  960, shanti nagar, road no.3,railway station d...  70.821817   9.820031   
7971  lavender 207 mangalam aananda sanganer,mansaro...  71.051058  11.394790   
1864  b-53b, anita colony road no. - 3, bajaj nagar,...  70.630176   9.266650   
8816  26 dev nager sanganer railway station jaipur ,...  70.985133  11.441609   
2388  plot no.61,sewaram nagar,near kanakpura railwa...  71.827947   7.060267   
1653  adinath traders shop no.40 railway station roa...  70.944240  11.258171   
5268  house no -60 ,prem nagar , near jagatpura rail...  70.371650  10.454823   
7226  gulab nagar sanaganer railway station ke saman...  70.971156  11.187452   
4306  gulab nagar sanaganer railway station ke saman...  70.971156  11.187452   
4856  c-77 anand vihar,railway colony, vidhaya dhar ...  70.078017  10.640264   
9707  195/109,sector - 19 pratap nagar sanganer ,nea...  70.193137  11.843340   
7794  39 bajaj nagar enclave near gandhi nagar railw...  70.612924   9.274251   
1137  bright cotton 2nd floor, shop no. 6 krishna pl...  70.844079  11.210822   
8820  p.no4 \n ambeshwer colony near shyam nagar met...  70.989298   8.340999   
5045  119 shiv colony new sanganer road sodala, near...  70.971027   8.321037   
1438  108 jasmine block manglam ananda #near sangane...  71.050630  11.357928   
...                                                 ...        ...        ...   
4173  c 1/32 chitrakootbehind global heart hospital,...  71.382938   7.994557   
4175  first crossing khawas ji ka rasta,hawa mahal r...  70.281734   7.227152   
4214  c-19a, jhalana doongari dharam kaante wali gal...  70.352478   9.321617   
4176  124/158 agarwal farm thari market mansover,nea...  71.009840  10.282056   
4213                         2/431,malviya nagar,302017  70.402334  10.078426   
4210               riet college,bhankrota,jaipur,302026  71.947078   8.982923   
4209               52/155,shipra path mansarover,302020  71.052184   9.473978   
4207  2nd floor central tower ,b-12 central spine ro...  70.854165   5.802891   
4206  rajasthan pathya pustak mandal,near pollution ...  70.459161   9.125567   
4205             124/80 ,thadi market mansarover,302020  71.014757  10.340170   
4203  plot no. 14, nirmal vihar, near kaushik school...  71.299065   5.551413   
4202  b 405 mahima desire jaisighpura bhankrota ajme...  71.865763   9.319644   
4201  11a, teacher's colony,,behind dcm, ajmer road,...  71.287653   8.377659   
4198  plot no: 05, shekhawat clinic basement, , shar...  70.825099   8.338593   
4197  122/111 vijay path , agarwal farm mansarovar,3...  71.015315  10.149435   
4196  apple society, flat- 69, floor-1, plot- a2-69,...  71.453669   8.090832   
4195  village thikariya nehar andhra bank,,infotech ...  72.821910  10.087185   
4194  52, 53 ashok vehar ambabari bridge jhotwara ro...  70.865885   6.766066  

In [16]:
df_copy.sort_values(by='distance', ascending=True)

address   latitude  longitude  \
0     jaipur,h.no.- 408, lavender mangalam, anand na...  71.052021  11.396546   
5201  jaipur,h.no.- 408, lavender mangalam, anand na...  71.051969  11.398205   
7971  lavender 207 mangalam aananda sanganer,mansaro...  71.051058  11.394790   
6903   104, lavendar,, mangalam aananda,sanganer,302029  71.053568  11.394492   
2076  104, lavendar,mangalam's aananda, sanganer,in ...  71.052389  11.403921   
7965  sunflower 410, manglam aananda,near rampura ro...  71.064959  11.394044   
3111  211 sunflower, manglam aananda city,opposite s...  71.064646  11.392212   
8866  flat no. 304,sun flower,manglam aananda, near ...  71.065814  11.411296   
6156  indrajeet singh , apollo pharmacy ,mangalam an...  71.071097  11.406526   
6907  g-03, orchid ,mangalam ananda ,road sanganer,n...  71.071004  11.412793   
5008          lotus 101,,aananda city, sanganer,,302029  71.040372  11.374064   
7800  lily,\n211, managalam ananda society,\nnew san...  71.064569  11.372417   
4435  g-02 orchid manglam ananda city ,rampur road n...  71.068659  11.421490   
9265  lily 303, manglam's ananda,rampura road, sanga...  71.066179  11.363584   
8961  flat no. 309  jasmine block  mangalam ananda ....  71.052083  11.360531   
7369  rose g-8 , mangalam aananda, opposite sanganer...  71.058382  11.359776   
8118  anshuman kumar, jasmine -g08, manglams aananda...  71.053519  11.359216   
5702  flat 306, block iris, mangalam ananda,near san...  71.045985  11.359348   
1438  108 jasmine block manglam ananda #near sangane...  71.050630  11.357928   
7974  rose g-8 , mangalam aananda, opposite sanganer...  71.062842  11.351647   
2992  house no.105,shayam vihar near aanada city san...  71.068248  11.349194   
8244  41 sunder nagar sanganer railway station, ke s...  70.999621  11.356692   
8816  26 dev nager sanganer railway station jaipur ,...  70.985133  11.441609   
7837  72, dev nagar-b sanganer railway station ke sa...  71.000879  11.331571   
7506  p.no. 98,dev nagar a hajyawala sanganer ( aana...  71.055892  11.485187   
5838  near railway station sanganer kankho ki dhani|...  70.967692  11.427949   
1573   91 , gulab nagar railway station sanganer,302020  70.973082  11.315466   
3287       b-70 dev nagar rampura road, sanganer,302029  71.003355  11.293942   
2235    plot 67(old 442/19),gopi nagar, sanganer,302029  70.935306  11.387877   
2112  h. no -6 heerawat nagar , sanganer , near opp ...  70.951640  11.328842   
...                                                 ...        ...        ...   
1620  hotel indana palace\njaipur - delhi highway, a...  69.671316   3.965697   
348   hotel indana palace\njaipur - delhi highway, a...  69.671316   3.965697   
4762  dr. mritunjya bhardwaj vatica,near govt. schoo...  71.788638   3.800507   
6168  vivek pharmachem india limited ,nh8 ,kunda che...  69.669496   3.843139   
3206  shankara institute of technolgy,natata kukas b...  69.480318   3.428038   
3796  jec kukas riico industrial area kukas ,jaipur,...  69.491426   3.413143   
5374  jaipur engineering college ,sp 43ricco industr...  69.495787   3.407108   
6448  jaipur engineering college ,sp 43ricco industr...  69.495787   3.407108   
5258  sp42 ,arya college of engineering & it jaipur,...  69.522432   3.385663   
2323  arya college of engineering and it ,kukas jaip...  69.516153   3.383443   
857    jw marriott jaipur, resort and spa kookes,302028  69.506160   3.385207   
3885  a1 hostel  arya college of engineering and it,...  69.519810   3.379794   
9482  arya vocation courses pizza wali building. sp-...  69.520693   3.378431   
9455        arya college ,near le meridien hotel,302028  69.521076   3.378255   
310   aceit kukass jaipur,ricco industrial area kuka...  69.516701   3.378119   
9806  arya college of engineering and it,riico, kuka...  69.519866   3.377343   
2074  arya college of engineering and information ,r...  69.521280   3.376215   
368   sp-42 riico industrial aera kukas jaipur,arya ...  69.519880   3.373856  

## Inferences to be drawn from the two tables :

** Jaccord Similarity and Distance between to addresses have a seemingly inverse relationship, which is in line with intuition as the more similar an address is the closer it should be**

**Anomalies in DATA: 
  There are various repetions of same addresses in our database with slight latitude/longitude differences which can be      attributed to GPS noise. We could potentially aggregate our location metrics for these repeating entries **
 
** Paying close attention to the table above element indexed '7965' has a similarity score of '0' but has the 5th lowest distance seperation when rank ordered based on distance, this is counterintuitive as low distance seperation should co-relate with high similarity. 
This can be explained by the fact that element indexed '7965' has tokens that are mispelled versions of tokens from our comparision set (eg. Manglam/Mangalam , Anand/Ananda...). Transforming to mitigate these spelling errors can definitely increase our similarity scores. **

 


# Prediction

In [17]:
print('input address')
address = str(input()) #Input new address

input address
JAIPUR,H.NO.- 408, LAVENDER MANGALAM, ANAND NAGAR SANGANER, RAILWAY STATION, JAIPUR - 302029 ,JAIPUR H O-302029 ,Rajasthan INDIA,302029


In [18]:
address= address.lower()
tokens = nltk.word_tokenize(address)
exceptions = [',','jaipur','rajasthan' ,'near', 'road', 'apartment', 'adjacent','sector']
tokens_final=set(tokens).difference(exceptions)
tokens_final

{'-',
 '302029',
 '408',
 'anand',
 'h',
 'h.no.-',
 'india,302029',
 'lavender',
 'mangalam',
 'nagar',
 'o-302029',
 'railway',
 'sanganer',
 'station'}

In [19]:
for i in range(len(df_copy)):
    intersection = tokens_final.intersection(df_copy['tokens'].iloc[i])
    jaccard_score= float(len(intersection))/((len(df_copy.tokens.iloc[0])) + len(df_copy.tokens.iloc[i]) - len(intersection))
    df_copy['similarity'].iloc[i] = jaccard_score

In [20]:
df_copy.sort_values(by='similarity', ascending=False).iloc[0].address #Nearest address based on only the address input

'jaipur,h.no.- 408, lavender mangalam, anand nagar sanganer, railway station, jaipur - 302029 ,jaipur h o-302029 ,rajasthan india,302029'